### PR TITLE
Update/M404

### PR DESCRIPTION
**Pull request by TamaroWalter**
WIP: change of format_string() and format_text() still need adjustments

### CODE REVIEW

File: lib.php [hunk 1]
```diff
@@ -883,8 +885,7 @@ function moodleoverflow_send_mails() {
 
                 // Preapare to actually send the post now. Build up the content.
                 $cleanname = str_replace('"', "'", strip_tags(format_string($moodleoverflow->name)));
-                $coursecontext = context_course::instance($course->id);
-                $shortname = format_string($course->shortname, true, ['context' => $coursecontext]);
+                $shortname = format_string($course->shortname, true, ['context' => context_course::instance($course->id)]);
```
Comment (NinaHerrmann) [inline lib.php:888]: Here and previously you still pass an object to the format_string function. 
> The $options array must not be a \context or \core\context object. This was never supported and behaviour was previously undetected

I guess you are supposed to just pass the name, but please check! (Or correct me if I misunderstood it!)

Comment (TamaroWalter) [reply lib.php:888]: As I understand it (but I'm a little bit confused) the array itself should not be an context-object (like function(param1, param2, object) ) but inside the array an object is okay ( like function(param1, param2, array('key' => contextobject)) )

File: .github/workflows/moodle-ci.yml [hunk 2]
```diff
@@ -8,8 +8,8 @@ jobs:
 
     strategy:
       matrix:
-        php: ['8.1']
-        moodle-branch: ['MOODLE_402_STABLE']
+        php: ['8.2']
+        moodle-branch: ['MOODLE_404_STABLE']
```
Comment (NinaHerrmann) [inline .github/workflows/moodle-ci.yml:12]: This branch does not exist yet you need to test against master

File: .github/workflows/moodle-ci.yml [hunk 3]
```diff
@@ -114,24 +114,6 @@ jobs:
         moodle-branch: ['MOODLE_401_STABLE', 'MOODLE_402_STABLE']
```
[the remark this answers is not in the document]
Comment (TamaroWalter) [reply .github/workflows/moodle-ci.yml]: done

[the remark this answers is not in the document]
Comment (NinaHerrmann) [reply .github/workflows/moodle-ci.yml]: please add the main branch as there is no 404_STABLE yet... :joy:

File: lib.php (hunk 1 above)
Comment (NinaHerrmann) [reply lib.php:888]: Hm, I looked again at it and I think it is illegal to do sth like this: 
`format_string($course->shortname, true, context_course::instance($course->id));`
(see function) [https://github.com/moodle/moodle/blob/757be30c3903bed55ad4a3607346b23510741745/lib/weblib.php#L1468]
Therefore, I guess the previous way was also okay just passing without making it an array is illegal.

File: .github/workflows/moodle-ci.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (NinaHerrmann) [reply .github/workflows/moodle-ci.yml]: You might want to look at [Opencast CI](https://github.com/Opencast-Moodle/moodle-workflows-opencast/blob/update/m404/.github/workflows/moodle-ci.yml)

File: lib.php (hunk 1 above)
Comment (TamaroWalter) [reply lib.php:888]: okay, im changing it back as it was

File: locallib.php [hunk 4]
```diff
@@ -450,10 +449,25 @@ function moodleoverflow_print_forum_list($course, $cm, $movetopopup) {
     $forumarray = [[]];
     $currentforum = $DB->get_record('moodleoverflow_discussions', ['id' => $movetopopup], 'moodleoverflow');
     $currentdiscussion = $DB->get_record('moodleoverflow_discussions', ['id' => $movetopopup], 'name');
-    $forums = $DB->get_records('moodleoverflow', ['course' => $course->id]);
+
+    // If the currentforum is anonymous, only show forums that have a higher anonymous setting.
+    $anonymoussetting = $DB->get_field('moodleoverflow', 'anonymous', ['id' => $currentforum->moodleoverflow]);
+    if ($anonymoussetting == anonymous::QUESTION_ANONYMOUS || $anonymoussetting == anonymous::EVERYTHING_ANONYMOUS) {
```
Comment (NinaHerrmann) [inline locallib.php:455]: It should be possible to move a post from a forum where questions are anonymous to a forum where questions are anonymous correct? Currently only everything anonymous forums are displayed.

File: tests/phpunit.xml
Comment (NinaHerrmann) [inline tests/phpunit.xml:1]: please remove the file git rm testunit.xml

File: locallib.php (hunk 4 above)
Comment (TamaroWalter) [reply locallib.php:455]: Yes it should be this way, it's fixed now.

**Pull request merged**